judo_cpp/rollout.cpp

```diff
@@ -0,0 +1,271 @@
+#include <omp.h>
+#include <string>
+
+namespace py = pybind11;
+
+// ========== Thread Pool Implementation ==========
+
+PersistentThreadPool::PersistentThreadPool(int num_threads)
+    : num_threads_(num_threads), stop_(false), active_workers_(0), total_tasks_(0), completed_tasks_(0) {
+    threads_.reserve(num_threads);
+    for (int i = 0; i < num_threads; ++i) {
+        threads_.emplace_back(&PersistentThreadPool::worker_thread, this);
+    }
+}
+
+PersistentThreadPool::~PersistentThreadPool() {
+    {
+        std::unique_lock<std::mutex> lock(queue_mutex_);
+        stop_ = true;
+    }
+    condition_.notify_all();
+    for (std::thread& thread : threads_) {
+        if (thread.joinable()) {
+            thread.join();
+        }
+    }
+}
+
+void PersistentThreadPool::worker_thread() {
+    while (true) {
+        std::function<void()> task;
+        {
+            std::unique_lock<std::mutex> lock(queue_mutex_);
+            condition_.wait(lock, [this] { return stop_ || !tasks_.empty(); });
+
+            if (stop_ && tasks_.empty()) {
+                return;
+            }
+
+            if (!tasks_.empty()) {
+                task = std::move(tasks_.front());
+                tasks_.pop();
+                active_workers_++;
+            }
+        }
+
+        if (task) {
+            task();
+
+            std::unique_lock<std::mutex> lock(queue_mutex_);
+            active_workers_--;
+            completed_tasks_++;
+            if (completed_tasks_ == total_tasks_) {
+                finished_.notify_one();
+            }
+        }
+    }
+}
+
+void PersistentThreadPool::execute_parallel(std::function<void(int)> func, int total_work) {
+    {
+        std::unique_lock<std::mutex> lock(queue_mutex_);
+        total_tasks_ = total_work;
+        completed_tasks_ = 0;
+
+        for (int i = 0; i < total_work; ++i) {
+            tasks_.emplace([func, i] { func(i); });
+        }
+    }
+
+    condition_.notify_all();
+
+    // Wait for all tasks to complete
+    std::unique_lock<std::mutex> lock(queue_mutex_);
+    finished_.wait(lock, [this] { return completed_tasks_ == total_tasks_; });
+}
+
+// Thread Pool Manager Implementation
+ThreadPoolManager& ThreadPoolManager::instance() {
+    static ThreadPoolManager instance;
+    return instance;
+}
+
+PersistentThreadPool* ThreadPoolManager::get_pool(int num_threads) {
+    std::lock_guard<std::mutex> lock(pool_mutex_);
+
+    if (!current_pool_ || current_num_threads_ != num_threads) {
+        current_pool_.reset();  // Destroy old pool
+        current_pool_ = std::make_unique<PersistentThreadPool>(num_threads);
+        current_num_threads_ = num_threads;
+    }
+
+    return current_pool_.get();
+}
+
+void ThreadPoolManager::shutdown() {
+    std::lock_guard<std::mutex> lock(pool_mutex_);
+    current_pool_.reset();
+    current_num_threads_ = 0;
+}
+
+// ========== End Thread Pool Implementation ==========
+
+py::array_t<double> make_array(std::vector<double>& buf,
```

Comment:
If you wish to move, then change signature to
```suggestion
py::array_t<double> make_array(std::vector<double>&& buf,
```